### DISCUSSION TITLE
Apriori Algorithm

### DISCUSSION THREAD

**Discussion by rafiff23** [teamalgoritma/community#461] [Q&A] [answered]
Saya mempunyai tabel sebagai berikut dengan minimum support yang diberikan adalah 2 dan confidence level = 50%

![image](https://user-images.githubusercontent.com/90491715/203159391-52322dbf-abe8-415e-afce-2558cc35581c.png)
![image](https://user-images.githubusercontent.com/90491715/203159316-b7f10a83-7477-4a0b-83b3-91c79ed369b1.png)

yang saya tangkap dari tabel tersebut adalah semua item yang ada itu termasuk _Frequent Item_ maka dari itu saya pertanyaan saya adalah bagaimana caranya membuat association rule nya?

**Comment by VictorNugraha** [comment]
Dear @rafiff23 ,

Terima kasih atas pertanyaan yang disampaikan, pertanyaan Anda sudah kami terima. Kami akan mencoba membantu mencarikan refrensi jawaban untuk pertanyaan yang diberikan.

**Comment by hwulanayu** [comment, marked as answer]
Dear @rafiff23,  tidak semua set yang ada di gambar termasuk frequent itemsets karena ada yang supportnya kurang dari 2. Jawaban singkat: hitung tiap confidence rule dari tiap set yang dapat dihasilkan dari frequent itemsets yang supportnya >= 2. jika hasil confidencenya >= 50% maka dapat dimasukan sebagai Association Rule.

Jawaban detail:

Diketahui:
TID: 20 transaksi
*minsup*: 2
*minconf*: 50%

TID:
l1,l2,l3
l3,l4,l2
l3,l4,l5
l2,l3,l1
l4,l5
l2,l4
l5,l1,l4
l3,l1
l5,l3
l1,l5,l4
l3,l1,l4
l4,l2,l1
l3,l2
l4,l5
l2,l3,l1
l2,l5
l3,l5,l4
l1,l3
l4,l1
l1,l5,l2

Universe U: {l1, l2, l3, l4, l5}

Association Rule: rule yang memiliki minimum support tertentu dan confidence minimum tertentu

Apa yang dimaksud dengan confidence? misal $X$ dan $Y$ adalah itemset yang berbeda. confidence dari rule $X => Y$  adalah kemungkinan atau probability sebuah transaksi yang mengandung itemset $Y$ terjadi, dengan kondisi transaksi tersebut mengandung itemset $X$.

Rumus: $\dfrac{sup(X \cup Y)}{sup(X)}$

contoh: $conf({Milk, Butter} => {Bread})$ adalah kemungkinan orang membeli Bread ketika orang tersebut sudah membeli Milk dan Butter.

Salah satu cara mencari Association Rule adalah dengan menggunakan algoritma Apriori yang komputasinya lebih kecil dibanding brute force $(2^n)$ dengan $n$ sama dengan jumlah transaksi.

Step-step mencari Association Rule:
1. Generate semua frequent itemsets dengan minimum support *minsup* = 2
2. Bentuk Association rule dari itemsets di step 1 dengan minimum confidence level sebesar *minconf* = 50%

Menggunakan algoritma apriori:

1. Generate semua frequent itemsets at minimum support 2/20:

apriori algorithm:
F1 =  l1, l2, l3, l4, l5
sup(l1) = 11/20
sup(l2) = 9/20
sup(l3) = 10/20
sup(l4) = 11/20
sup(l5) =  9/20

C2 (before pruning): (Generate set dari F1 dan cek tiap subset di bawah ada gak di F1. contoh: set {l1,l2} mempunyai subset l1 dan l2. keduanya ada di F1 maka tidak dipruning)
{l1,12}
{11, 13}
{11, 14}
{11, 15}
{12, 13}
{12, 14}
{12, 15}
{13, 14}
{13, 15}
{14, 15}

C2 (after pruning) = C2 (before pruning) (tidak ada yg diprune, semua subset ada semua di F1)
hitung tiap supportnya, minimum support adalah 2. tiap set yg supportnya kurang dari 2, drop.
{l1,12} = 5/20
{11, 13} = 5/20
{11, 14} = 4/20
{11, 15} = 3/20
{12, 13} = 5/20
{12, 14} = 3/20
{12, 15} = 2/20
{13, 14} = 4/20
{13, 15} = 2/20
{14, 15} = 6/20

F2: (hasil dari C2 dengan minsup = 2)
{l1,l2}
{l1, l3}
{l1, l4}
{l1, l5}
{l2, l3}
{l2, l4}
{l2, l5}
{l3, l4}
{l3, l5}
{l4, l5}

C3 (before pruning): (Generate set dari F2 dan cek tiap subset di bawah ada gak di F2. contoh: set {l1,l2,l3} mempunyai subset {l1,l2}
{l1, l3} dan {l2, l3}. ketiganya ada di F2 maka tidak dipruning)
{l1, l2, l3}
{l1, l2, l4}
{l1, l2, l5}
{l1, l4, l5}
{l1, l4, l3}
{l1, l5, l3}
{l2, l3, l5}
{l2, l4, l5}
{l2, l4, l3}
{l4, l5, l3}


C3 (after pruning) = C3 (before pruning) (tidak ada yg diprune, semua subset ada semua di F2)
hitung tiap supportnya, minimum support adalah 2. tiap set yg supportnya kurang dari 2, drop.
{l1, l2, l3} = 3
{l1, l2, l4} = 1 DROP
{l1, l2, l5} = 1 DROP
{l1, l4, l5} = 2
{l1, l4, l3} = 1 DROP
{l1, l5, l3} = 0 DROP
{l2, l3, l5} = 0 DROP
{l2, l4, l5} = 0 DROP
{l2, l4, l3} = 1 DROP
{l4, l5, l3} = 2

F3: (hasil dari C3 dengan minsup = 2)
{l1, l2, l3} = 3
{l1, l4, l5} = 2
{l4, l5, l3} = 2

C4 (before pruning) =
{l1, l3, l4, l5} PRUNE karena tidak ada subset {l1, l3, l4} dan {l1, l3, l5} di F3

C4 (after pruning) = -
F4 = -

Algoritma berhenti

Frequent pattern: F1 U F2 U F3 U F4:
{l1}
{l2}
{l3}
{l4}
{l5}
{l1,l2}
{l1, l3}
{l1, l4}
{l1, l5}
{l2, l3}
{l2, l4}
{l2, l5}
{l3, l4}
{l3, l5}
{l4, l5}
{l1, l2, l3}
{l1, l4, l5}
{l4, l5, l3}

2. Association Rule:

hitung confidence tiap set di atas dengan minconf = 50% menggunakan rumus: 
$\dfrac{sup(X \cup Y)}{sup(X)}$

contoh: 
$conf(l3 => l1) = \dfrac{sup(l3 \cup l1)}{sup(l3)}$ = 5/20 / 10/20  = 1/2 = 50% <---- masuk Association Rule

Jangan lupa pertimbangkan Confidence monotonicity property dan hilangkan rule yang lebih kecil.

Jawaban Association Rule lebih lengkapnya dapat dilihat di  [file ini](https://drive.google.com/file/d/1fiKge9W9HlQ_Lli1mwZBtm759v7F1ui3/view?usp=sharing) dengan membuka lewat aplikasi [Orange Data Mining](https://orangedatamining.com/).

**Reply by rafif24** [reply]
Terimakasih Babnyak ka!

**Reply by rafif24** [reply]
untuk contoh yang dipakai itu itemset yang {I3, I1} ?

**Reply by rafif24** [reply]
apakah bisa diberiknan contoh dari itemset {I1}, {I1,I2}, dan {I1, I2, I3}

**Reply by hwulanayu** [reply]
Dear @rafiff23,
 
Bukan, yang dipakai adalah itemset {l3} dan {l1}. 

Jika ingin membuat confidence rule dengan itemsets {I1}, {I1,I2}, dan {I1, I2, I3}, maka akan **menyalahi** aturan bahwa $X$ dan $Y$ dalam $conf(X => Y)$ adalah himpunan atau set **disjoint**. X dan Y tidak boleh memiliki elemen yang sama. 

Sekedar contoh perhitungan itemset {l1, l2} dengan {l1}:

$conf({l1, l2} => {l1}) =  \dfrac{sup({l1, l2} \cup {l1})}{sup({l1})} = \dfrac{sup(l1, l2)}{sup(l1)} = \dfrac{\dfrac{5}{20}}{\dfrac{11}{20}} = 0.454545454545$ = 45% <----- tidak boleh digunakan untuk association rule karena menyalahi aturan $X \cap Y = \emptyset$

**Reply by rafif24** [reply]
ohhh gitu oke makasih ya ka
